File: 12-extended-dll/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Struktur Node untuk Extended Double Linked List
typedef struct DLLNode {
    int data;
    struct DLLNode* prev;
    struct DLLNode* next;
} DLLNode;

// Struktur untuk mengelola Extended Double Linked List
typedef struct {
    DLLNode* head;
    DLLNode* tail;
    int size;
} DoublyLinkedList;

// Fungsi untuk inisialisasi list
DoublyLinkedList* createList() {
    DoublyLinkedList* list = (DoublyLinkedList*)malloc(sizeof(DoublyLinkedList));
    if (list == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
    return list;
}

// Fungsi untuk membuat node baru
DLLNode* createNode(int data) {
    DLLNode* newNode = (DLLNode*)malloc(sizeof(DLLNode));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

// Fungsi untuk menambah node di awal
void insertAtBeginning(DoublyLinkedList* list, int data) {
    DLLNode* newNode = createNode(data);
    
    if (list->head == NULL) {
        list->head = newNode;
        list->tail = newNode;
    } else {
        newNode->next = list->head;
        list->head->prev = newNode;
        list->head = newNode;
    }
    
    list->size++;
}

// Fungsi untuk menambah node di akhir
void insertAtEnd(DoublyLinkedList* list, int data) {
    DLLNode* newNode = createNode(data);
    
    if (list->tail == NULL) {
        list->head = newNode;
        list->tail = newNode;
    } else {
        newNode->prev = list->tail;
        list->tail->next = newNode;
        list->tail = newNode;
    }
    
    list->size++;
}

// Fungsi untuk menambah node di posisi tertentu
void insertAtPosition(DoublyLinkedList* list, int data, int position) {
    if (position < 0 || position > list->size) {
        printf("Invalid position\n");
        return;
    }
    
    if (position == 0) {
        insertAtBeginning(list, data);
        return;
    }
    
    if (position == list->size) {
        insertAtEnd(list, data);
        return;
    }
    
    DLLNode* newNode = createNode(data);
    DLLNode* current = list->head;
    
    for (int i = 0; i < position - 1; i++) {
        current = current->next;
    }
    
    newNode->next = current->next;
    newNode->prev = current;
    current->next->prev = newNode;
    current->next = newNode;
    
    list->size++;
}

// Fungsi untuk mencari node
DLLNode* search(DoublyLinkedList* list, int key) {
    DLLNode* current = list->head;
    
    while (current != NULL) {
        if (current->data == key) {
            return current;
        }
        current = current->next;
    }
    
    return NULL;
}

// Fungsi untuk menghapus node di awal
void deleteAtBeginning(DoublyLinkedList* list) {
    if (list->head == NULL) {
        return;
    }
    
    DLLNode* temp = list->head;
    
    if (list->head == list->tail) {
        list->head = NULL;
        list->tail = NULL;
    } else {
        list->head = list->head->next;
        list->head->prev = NULL;
    }
    
    free(temp);
    list->size--;
}

// Fungsi untuk menghapus node di akhir
void deleteAtEnd(DoublyLinkedList* list) {
    if (list->tail == NULL) {
        return;
    }
    
    DLLNode* temp = list->tail;
    
    if (list->head == list->tail) {
        list->head = NULL;
        list->tail = NULL;
    } else {
        list->tail = list->tail->prev;
        list->tail->next = NULL;
    }
    
    free(temp);
    list->size--;
}

// Fungsi untuk menghapus node berdasarkan key
int deleteNode(DoublyLinkedList* list, int key) {
    if (list->head == NULL) {
        return 0;
    }
    
    // Jika node yang dihapus adalah head
    if (list->head->data == key) {
        deleteAtBeginning(list);
        return 1;
    }
    
    // Jika node yang dihapus adalah tail
    if (list->tail->data == key) {
        deleteAtEnd(list);
        return 1;
    }
    
    // Mencari node di tengah
    DLLNode* current = list->head->next;
    
    while (current != list->tail) {
        if (current->data == key) {
            current->prev->next = current->next;
            current->next->prev = current->prev;
            free(current);
            list->size--;
            return 1;
        }
        current = current->next;
    }
    
    return 0; // Key tidak ditemukan
}

// Fungsi untuk memutar list (reverse)
void reverseList(DoublyLinkedList* list) {
    if (list->head == NULL || list->head == list->tail) {
        return;
    }
    
    DLLNode* current = list->head;
    DLLNode* temp = NULL;
    
    // Menukar prev dan next untuk semua node
    while (current != NULL) {
        temp = current->prev;
        current->prev = current->next;
        current->next = temp;
        current = current->prev;
    }
    
    // Menukar head dan tail
    temp = list->head;
    list->head = list->tail;
    list->tail = temp;
}

// Fungsi untuk menampilkan list (forward)
void displayForward(DoublyLinkedList* list) {
    if (list->head == NULL) {
        printf("List is empty\n");
        return;
    }
    
    DLLNode* current = list->head;
    printf("List (forward): ");
    
    while (current != NULL) {
        printf("%d ", current->data);
        current = current->next;
    }
    
    printf("\n");
}

// Fungsi untuk menampilkan list (backward)
void displayBackward(DoublyLinkedList* list) {
    if (list->tail == NULL) {
        printf("List is empty\n");
        return;
    }
    
    DLLNode* current = list->tail;
    printf("List (backward): ");
    
    while (current != NULL) {
        printf("%d ", current->data);
        current = current->prev;
    }
    
    printf("\n");
}

// Fungsi untuk membersihkan list
void freeList(DoublyLinkedList* list) {
    DLLNode* current = list->head;
    DLLNode* next;
    
    while (current != NULL) {
        next = current->next;
        free(current);
        current = next;
    }
    
    free(list);
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    // Inisialisasi double linked list
    DoublyLinkedList* list = createList();
    
    start = clock();
    
    // Operasi pada double linked list
    // Menambahkan node di awal dan akhir
    for (int i = 0; i < 400; i++) {
        insertAtBeginning(list, i);
    }
    
    for (int i = 400; i < 800; i++) {
        insertAtEnd(list, i);
    }
    
    // Menambahkan node di posisi tertentu
    for (int i = 800; i < 1000; i++) {
        int position = i % list->size;
        insertAtPosition(list, i, position);
    }
    
    // Mencari beberapa nilai
    for (int i = 0; i < 100; i++) {
        int key = (i * 17) % 1000;
        DLLNode* result = search(list, key);
        if (result != NULL) {
            // printf("Found %d\n", key);
        } else {
            // printf("%d not found\n", key);
        }
    }
    
    // Menghapus beberapa node
    for (int i = 0; i < 50; i++) {
        deleteAtBeginning(list);
    }
    
    for (int i = 0; i < 50; i++) {
        deleteAtEnd(list);
    }
    
    for (int i = 0; i < 100; i++) {
        int key = (i * 19) % 1000;
        deleteNode(list, key);
    }
    
    // Memutar list
    reverseList(list);
    
    // Menampilkan list (opsional, dimatikan untuk pengukuran performa)
    // displayForward(list);
    // displayBackward(list);
    
    // Membersihkan memori
    freeList(list);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Waktu eksekusi: %f detik\n", cpu_time_used);
    
    return 0;
}

File: 12-extended-dll/code.rs
use std::time::Instant;
use std::rc::{Rc, Weak};
use std::cell::RefCell;

// Definisi tipe untuk node references
type NodeRef<T> = Rc<RefCell<Node<T>>>;
type WeakNodeRef<T> = Weak<RefCell<Node<T>>>;

// Struktur Node untuk Extended Double Linked List
struct Node<T> {
    data: T,
    prev: Option<WeakNodeRef<T>>,
    next: Option<NodeRef<T>>,
}

// Implementasi untuk Node
impl<T> Node<T> {
    fn new(data: T) -> Self {
        Node {
            data,
            prev: None,
            next: None,
        }
    }
}

// Struktur untuk mengelola Extended Double Linked List
struct DoublyLinkedList<T> {
    head: Option<NodeRef<T>>,
    tail: Option<WeakNodeRef<T>>,
    size: usize,
}

// Implementasi untuk DoublyLinkedList
impl<T> DoublyLinkedList<T> 
where T: PartialEq + Copy {
    // Membuat double linked list baru
    fn new() -> Self {
        DoublyLinkedList {
            head: None,
            tail: None,
            size: 0,
        }
    }
    
    // Menambahkan node di awal
    fn insert_at_beginning(&mut self, data: T) {
        let new_node = Rc::new(RefCell::new(Node::new(data)));
        
        match self.head.take() {
            Some(old_head) => {
                old_head.borrow_mut().prev = Some(Rc::downgrade(&new_node));
                new_node.borrow_mut().next = Some(old_head);
                self.head = Some(new_node);
            },
            None => {
                self.tail = Some(Rc::downgrade(&new_node));
                self.head = Some(new_node);
            }
        }
        
        self.size += 1;
    }
    
    // Menambahkan node di akhir
    fn insert_at_end(&mut self, data: T) {
        let new_node = Rc::new(RefCell::new(Node::new(data)));
        
        match self.tail.take() {
            Some(old_tail) => {
                if let Some(old_tail_strong) = old_tail.upgrade() {
                    old_tail_strong.borrow_mut().next = Some(new_node.clone());
                    new_node.borrow_mut().prev = Some(Rc::downgrade(&old_tail_strong));
                    self.tail = Some(Rc::downgrade(&new_node));
                }
            },
            None => {
                self.head = Some(new_node.clone());
                self.tail = Some(Rc::downgrade(&new_node));
            }
        }
        
        self.size += 1;
    }
    
    // Menambahkan node di posisi tertentu
    fn insert_at_position(&mut self, data: T, position: usize) -> bool {
        if position > self.size {
            return false;
        }
        
        if position == 0 {
            self.insert_at_beginning(data);
            return true;
        }
        
        if position == self.size {
            self.insert_at_end(data);
            return true;
        }
        
        let new_node = Rc::new(RefCell::new(Node::new(data)));
        let mut current = self.head.clone();
        
        for _ in 0..position-1 {
            if let Some(node) = current {
                // PERBAIKAN: Gunakan clone untuk next node
                let next = node.borrow().next.clone();
                current = next;
            }
        }
        
        if let Some(current_node) = current {
            let next_node = current_node.borrow().next.clone();
            
            if let Some(next) = next_node {
                new_node.borrow_mut().next = Some(next.clone());
                new_node.borrow_mut().prev = Some(Rc::downgrade(&current_node));
                
                next.borrow_mut().prev = Some(Rc::downgrade(&new_node));
                current_node.borrow_mut().next = Some(new_node);
                
                self.size += 1;
                return true;
            }
        }
        
        false
    }
    
    // Mencari node dalam linked list
    fn search(&self, key: T) -> bool {
        let mut current = self.head.clone();
        
        while let Some(node) = current {
            if node.borrow().data == key {
                return true;
            }
            // PERBAIKAN: Gunakan clone untuk next node
            let next = node.borrow().next.clone();
            current = next;
        }
        
        false
    }
    
    // Menghapus node di awal
    fn delete_at_beginning(&mut self) -> bool {
        if self.head.is_none() {
            return false;
        }
        
        let old_head = self.head.take().unwrap();
        
        match old_head.borrow_mut().next.take() {
            Some(new_head) => {
                new_head.borrow_mut().prev = None;
                self.head = Some(new_head);
            },
            None => {
                self.tail = None;
            }
        }
        
        self.size -= 1;
        true
    }
    
    // Menghapus node di akhir
    fn delete_at_end(&mut self) -> bool {
        if self.tail.is_none() {
            return false;
        }
        
        if let Some(tail) = self.tail.take() {
            if let Some(tail_node) = tail.upgrade() {
                // PERBAIKAN: Ambil prev dalam blok terpisah untuk menghindari borrow issue
                let prev_weak = {
                    let mut tail_ref = tail_node.borrow_mut();
                    tail_ref.prev.take()
                };
                
                match prev_weak {
                    Some(prev_weak) => {
                        if let Some(prev) = prev_weak.upgrade() {
                            prev.borrow_mut().next = None;
                            self.tail = Some(Rc::downgrade(&prev));
                        }
                    },
                    None => {
                        self.head = None;
                    }
                }
                
                self.size -= 1;
                return true;
            }
        }
        
        false
    }
    
    // Menghapus node berdasarkan key
    fn delete_node(&mut self, key: T) -> bool {
        if self.head.is_none() {
            return false;
        }
        
        // Jika node yang akan dihapus adalah head
        if let Some(head) = &self.head {
            if head.borrow().data == key {
                return self.delete_at_beginning();
            }
        }
        
        // Jika node yang akan dihapus adalah tail
        if let Some(tail_weak) = &self.tail {
            if let Some(tail) = tail_weak.upgrade() {
                if tail.borrow().data == key {
                    return self.delete_at_end();
                }
            }
        }
        
        // Mencari node di tengah
        let mut current = self.head.clone();
        
        while let Some(node) = current {
            if node.borrow().data == key {
                // Mengambil prev dan next nodes
                let prev_weak = node.borrow().prev.clone();
                let next_opt = node.borrow().next.clone();
                
                if let (Some(prev_weak), Some(next)) = (prev_weak, next_opt) {
                    if let Some(prev) = prev_weak.upgrade() {
                        // Update links
                        prev.borrow_mut().next = Some(next.clone());
                        next.borrow_mut().prev = Some(Rc::downgrade(&prev));
                        
                        self.size -= 1;
                        return true;
                    }
                }
            }
            
            // PERBAIKAN: Gunakan clone untuk next node
            let next = node.borrow().next.clone();
            current = next;
        }
        
        false
    }
    
    // Memutar list (reverse)
    fn reverse(&mut self) {
        if self.head.is_none() || self.size <= 1 {
            return;
        }
        
        let mut current = self.head.take();
        let mut prev: Option<NodeRef<T>> = None;
        
        self.tail = match &current {
            Some(node) => Some(Rc::downgrade(node)),
            None => None,
        };
        
        while let Some(current_node) = current {
            // Simpan next sebelum diubah
            let next = current_node.borrow_mut().next.take();
            
            // Tukar prev dan next
            current_node.borrow_mut().next = prev.clone();
            
            if let Some(prev_node) = &prev {
                current_node.borrow_mut().prev = Some(Rc::downgrade(prev_node));
            } else {
                current_node.borrow_mut().prev = None;
            }
            
            // Lanjut ke node berikutnya
            prev = Some(current_node);
            current = next;
        }
        
        // Update head
        self.head = prev;
    }
}

fn main() {
    let start = Instant::now();
    
    // Inisialisasi extended double linked list
    let mut list = DoublyLinkedList::new();
    
    // Menambahkan node di awal dan akhir
    for i in 0..400 {
        list.insert_at_beginning(i);
    }
    
    for i in 400..800 {
        list.insert_at_end(i);
    }
    
    // Menambahkan node di posisi tertentu
    for i in 800..1000 {
        let position = i % list.size;
        list.insert_at_position(i, position);
    }
    
    // Mencari beberapa nilai
    for i in 0..100 {
        let key = (i * 17) % 1000;
        let found = list.search(key);
        if found {
            // println!("Found {}", key);
        } else {
            // println!("{} not found", key);
        }
    }
    
    // Menghapus beberapa node
    for _ in 0..50 {
        list.delete_at_beginning();
    }
    
    for _ in 0..50 {
        list.delete_at_end();
    }
    
    for i in 0..100 {
        let key = (i * 19) % 1000;
        list.delete_node(key);
    }
    
    // Memutar list
    list.reverse();
    
    let duration = start.elapsed();
    println!("Waktu eksekusi: {:?}", duration);
}